[test] [tensorflow] [ec2] Change TF EC2 tests to be more frugal with instance spin-up

saimidu opened the pull request:
*Issue #, if available:*

## Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [x] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [x] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Reviewer Checklist
* For reviewer, before merging, please cross-check:
- [x] I've verified the code change on the config file mentioned above has already been reverted

*Description:*
- Changed instance type used for TF EC2 tests to g3.16xlarge
- Added pytest fixtures to allow skipping tests through fixture parametrization, rather than skip inside test function
- Added more TF EC2 tests to list of allowed single-gpu tests with review from @satishpasumarthi 

*Tests run:*
- Tested with g2.8xlarge instance type for TF EC2 tests, but there are test failures on:
  - test_tensorflow_training.py::test_tensorflow_keras_horovod_amp
  - test_tensorflow_training.py::test_tensorflow_standalone_gpu
  - test_tensorflow_training.py::test_tensorflow_train_mnist_gpu
  - test_tensorflow_training.py::test_tensorflow_keras_horovod_fp32
  - test_tensorflow_training.py::test_tensorflow_tensorboard_gpu
  - test_tensorflow_training.py::test_tensorflow_addons_gpu
- Tested with g3.4xlarge instance type for TF EC2 tests, and succeeded on all tests, but g3.4xlarge only has a single GPU, similar to p2.xlarge
- Tested with g3.16xlarge instance type, and succeeded on all tests

*DLC image/dockerfile:*

*Additional context:*


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -18,13 +18,13 @@
 TF_DATASERVICE_TEST_CMD = os.path.join(CONTAINER_TESTS_PREFIX, "testDataservice")
 
-TF_EC2_GPU_INSTANCE_TYPE = get_ec2_instance_type(default="p2.8xlarge", processor="gpu")
+TF_EC2_GPU_INSTANCE_TYPE = get_ec2_instance_type("g3.16xlarge", "gpu")
```
why remove keywords? it's nice for clarity imo

saimidu replied:
Reverted this change.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -39,7 +39,7 @@ def test_tensorflow_standalone_cpu(tensorflow_training, ec2_connection, cpu_only
 
 @pytest.mark.model("mnist")
-@pytest.mark.parametrize("ec2_instance_type", TF_EC2_GPU_INSTANCE_TYPE, indirect=True)
+@pytest.mark.parametrize("ec2_instance_type", get_ec2_instance_type("g3.16xlarge", "gpu"), indirect=True)
```
why can't this be TF_EC2_GPU_INSTANCE_TYPE?

saimidu replied:
Reverted this change.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -18,13 +18,13 @@
 @pytest.mark.integration('tensorflow_sanity_test')
 @pytest.mark.model("N/A")
-@pytest.mark.parametrize("ec2_instance_type", TF_EC2_GPU_INSTANCE_TYPE, indirect=True)
+@pytest.mark.parametrize("ec2_instance_type", get_ec2_instance_type("g3.16xlarge", "gpu"), indirect=True)
```
why can't this be TF_EC2_GPU_INSTANCE_TYPE?

saimidu replied:
Reverted this change.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -52,7 +52,7 @@ def test_tensorflow_train_mnist_cpu(tensorflow_training, ec2_connection, cpu_onl
 @pytest.mark.integration("horovod")
 @pytest.mark.model("resnet")
-@pytest.mark.parametrize("ec2_instance_type", TF_EC2_GPU_INSTANCE_TYPE, indirect=True)
+@pytest.mark.parametrize("ec2_instance_type", get_ec2_instance_type("p2.8xlarge", "gpu"), indirect=True)
```
Why is this p2.8x? if this needs to persist, then you can leave it this way, or define another constant above if many tests need to be p2.8x in PR context

saimidu replied:
The test was failing for g3.16xlarge, but it turned out that the `shm_large` flag was only enabled for p2.8xlarge instances. When I enabled it for g3.16xlarge as well, it ran successfully.

So this change has been reverted as well.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -69,7 +69,7 @@ def test_tensorflow_with_horovod_cpu(tensorflow_training, ec2_connection, cpu_on
 @pytest.mark.integration("opencv")
 @pytest.mark.model("unknown_model")
-@pytest.mark.parametrize("ec2_instance_type", TF_EC2_GPU_INSTANCE_TYPE, indirect=True)
+@pytest.mark.parametrize("ec2_instance_type", get_ec2_instance_type("g3.16xlarge", "gpu"), indirect=True)
```
why can't this be TF_EC2_GPU_INSTANCE_TYPE?

saimidu replied:
Reverted this change.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -105,7 +105,7 @@ def test_tensorflow_telemetry_cpu(tensorflow_training, ec2_connection, cpu_only)
 @pytest.mark.integration("keras, horovod, automatic_mixed_precision (AMP)")
 @pytest.mark.model("mnist")
-@pytest.mark.parametrize("ec2_instance_type", TF_EC2_GPU_INSTANCE_TYPE, indirect=True)
+@pytest.mark.parametrize("ec2_instance_type", get_ec2_instance_type("g3.16xlarge", "gpu"), indirect=True)
```
why can't this be TF_EC2_GPU_INSTANCE_TYPE?

saimidu replied:
Reverted this change.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -114,7 +114,7 @@ def test_tensorflow_keras_horovod_amp(tensorflow_training, ec2_connection, gpu_o
 @pytest.mark.integration("keras, horovod, single_precision_floating_point (FP32)")
 @pytest.mark.model("mnist")
-@pytest.mark.parametrize("ec2_instance_type", TF_EC2_GPU_INSTANCE_TYPE, indirect=True)
+@pytest.mark.parametrize("ec2_instance_type", get_ec2_instance_type("g3.16xlarge", "gpu"), indirect=True)
```
why can't this be TF_EC2_GPU_INSTANCE_TYPE?

saimidu replied:
Reverted this change.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -124,7 +124,7 @@ def test_tensorflow_keras_horovod_fp32(tensorflow_training, ec2_connection, gpu_
 @pytest.mark.integration("tensorboard, keras")
 @pytest.mark.model("sequential")
-@pytest.mark.parametrize("ec2_instance_type", TF_EC2_GPU_INSTANCE_TYPE, indirect=True)
+@pytest.mark.parametrize("ec2_instance_type", get_ec2_instance_type("g3.16xlarge", "gpu"), indirect=True)
```
why can't this be TF_EC2_GPU_INSTANCE_TYPE?

saimidu replied:
Reverted this change.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -145,7 +145,7 @@ def test_tensorflow_tensorboard_cpu(tensorflow_training, ec2_connection, cpu_onl
 @pytest.mark.model("sequential")
 @pytest.mark.integration("tensorflow_addons, keras")
-@pytest.mark.parametrize("ec2_instance_type", TF_EC2_GPU_INSTANCE_TYPE, indirect=True)
+@pytest.mark.parametrize("ec2_instance_type", get_ec2_instance_type("g3.16xlarge", "gpu"), indirect=True)
```
why can't this be TF_EC2_GPU_INSTANCE_TYPE?

saimidu replied:
Reverted this change.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -183,7 +183,7 @@ def test_tensorflow_dataservice_cpu(tensorflow_training, ec2_connection, cpu_onl
 @pytest.mark.integration('tensorflow-dataservice-test')
 @pytest.mark.model("N/A")
-@pytest.mark.parametrize("ec2_instance_type", TF_EC2_GPU_INSTANCE_TYPE, indirect=True)
+@pytest.mark.parametrize("ec2_instance_type", get_ec2_instance_type("g3.16xlarge", "gpu"), indirect=True)
```
why can't this be TF_EC2_GPU_INSTANCE_TYPE?

saimidu replied:
Reverted this change.

arjkesh commented on test/dlc_tests/ec2/tensorflow/training/test_tensorflow_training.py:
```diff
@@ -53,10 +54,14 @@ def test_tensorflow_train_mnist_cpu(tensorflow_training, ec2_connection, cpu_onl
+        connection=ec2_connection,
+        ecr_uri=tensorflow_training,
+        test_cmd=test_script,
+        large_shm=bool(re.match(r"(p2\.8xlarge)|(g3\.16xlarge)", ec2_instance_type))
```
Did you try this on different gpu instance types this test might run on? If so, can you update the tests run section?

saimidu replied:
Currently experimenting with g3.4xlarge. Will update as soon as those tests run.

arjkesh reviewed (commented):
Please update description to reflect g3.16 changes

satishpasumarthi reviewed (approved):
Add a comment on the test mentioning "for t2.3 and above etc." before function definition? It's hard to judge from the arguments.

satishpasumarthi reviewed (commented):
in test_tensorflow_training.py can we classify into 3 classes having tf2, tf1 and common tests?
i.e., maintain separate classes for all 3? something like: 
class TestTF2 {}
class TestTF1 {}
class TestAll {}

saimidu commented:
Thanks @satishpasumarthi for the comments. Addressed the first, and added better comments for the skipped test.

Regarding the refactoring of tests, will get around to this later as part of a future PR.

arjkesh commented on test/dlc_tests/conftest.py:
```diff
@@ -361,14 +377,22 @@ def pytest_generate_tests(metafunc):
                 if lookup in image:
                     is_example_lookup = "example_only" in metafunc.fixturenames and "example" in image
                     is_standard_lookup = "example_only" not in metafunc.fixturenames and "example" not in image
+                    tf2_requirement_failed = "tf2_only" in metafunc.fixturenames and is_tf1(image)
```
if we need to add more tf specific conditions like this, will we have to bloat this function even further? I suggest having a separate function to handle any tf2 specific requirements.

Additionally, why do we have functions like is_tf20 instead of having functions that are is_tf_version(image, tf_version) - in terms of scaling for future versions it becomes cumbersome to create a new function every time

saimidu replied:
I actually don't know why those were added. I can try to clean this whole set of fixtures up, and make the TF 2 fixtures generic.

arjkesh replied:
please focus on just cleaning up the tf logic that was added so as not to outscope this PR

arjkesh commented on test/dlc_tests/conftest.py:
```diff
@@ -276,6 +277,37 @@ def example_only():
+    :param image: Image URI for which the validation must be performed
+    :return: True if all validation succeeds, else False
+    """
+    tf2_requirement_failed = "tf2_only" in metafunc_obj.fixturenames and is_tf1(image)
```
nit: wouldn't it be more appropriate for this condition to be `"tf2_only" in metafunc_obj.fixturenames and not is_tf2(image)` - i.e. it's possible someone uses a fixture other than tensorflow_training or tensorflow_inference but also has this tf2_only fixture.

saimidu replied:
Sure. Alright.

saimidu replied:
Sure. That's true. I'll make the change.

arjkesh commented on test/dlc_tests/conftest.py:
```diff
@@ -276,6 +277,37 @@ def example_only():
+    """
+    tf2_requirement_failed = "tf2_only" in metafunc_obj.fixturenames and is_tf1(image)
+    tf23_requirement_failed = "tf23_and_above_only" in metafunc_obj.fixturenames and below_tf23(image)
+    tf21_requirement_failed = "tf21_and_above_only" in metafunc_obj.fixturenames and (is_tf1(image) or is_tf20(image))
```
tf23_requirement_failed and tf21_requirement_failed essentially are trying to accomplish the same thing with different tf versions, but are implemented differently. Let's unify this - can you create a general function for checking if tf version is below a certain threshold (i.e. `is_below_tf_version(image, tf_version)`) and call it in both cases?

saimidu replied:
Yeah, I was thinking of doing this, but didn't want to go out of scope on this PR. I already have an implementation for this. Will update and submit another revision.

arjkesh commented on test/test_utils/test_reporting.py:
```diff
@@ -43,8 +43,15 @@ class TestReportGenerator:
+        "test_framework_and_cuda_version_gpu",
+        "test_curand_gpu",
+        "test_dependency_check_gpu",
+        "test_tensorflow_addons_gpu",
+        "test_tensorflow_standalone_gpu",
+        "test_tensorflow_tensorboard_gpu",
```
@sboshin could you sign off that these tests are allowed to be single gpu?

sboshin replied:
test_tensorflow_tensorboard_gpu should be multi-gpu.
Everything else looks fine to me.

saimidu replied:
Thanks! Will make the change.

SergTogul reviewed (approved):
LGTM
